src/NuGet.Core/NuGet.Packaging/PackageReaderBase.cs

```diff
@@ -536,6 +536,46 @@ protected static string GetNuspecFile(IEnumerable<string> files)
             return nuspecPaths.Single();
         }
 
+        /// <summary>
+        /// Validate file entry in package is not traversed outside of the expected extraction path.
+        /// Eg: file entry like ../../foo.dll can get outside of the expected extraction path.
+        /// </summary>
+        protected static void ValidatePackageEntry(string normalizedDestination, string normalizedFilePath, PackageIdentity packageIdentity)
+        {
+            // Destination and filePath must be normalized.
+            var fullPath = Path.GetFullPath(Path.Combine(normalizedDestination, normalizedFilePath));
+
+            if(!fullPath.StartsWith(normalizedDestination, StringComparison.Ordinal) || fullPath.Length == normalizedDestination.Length)
+            {
+                throw new UnsafePackageEntryException(string.Format(
+                    CultureInfo.CurrentCulture,
+                    Strings.ErrorUnsafePackageEntry,
+                    packageIdentity.ToString()));
+            }
+        }
+
+        protected string NormalizesDirectoryPath(string path)
+        {
+            if (!path.EndsWith(Path.DirectorySeparatorChar.ToString()))
```

Comment:
You can consider `path[path.Length-1] != Path.DirectorySeparatorChar`, might be a bit faster than doing string comparison.

Reply:
EndsWith() actually did almost like that check